### PR TITLE
feat: add watermark to UI

### PR DESCRIPTION
**Pull request by Krithika3**
Let us decide if we like the placement here:
Screen Shot 2022-07-24 at 2 45 43 PM

And will not be displayed if isWatermark enabled is not true

### CODE REVIEW

File: server/routes/routes.go [hunk 1]
```diff
@@ -33,6 +33,6 @@ func v1Routes(r gin.IRouter) {
 	r.GET("/metrics/namespaces/:namespace/pods/:pod", handler.GetPodMetrics)
 	r.GET("/namespaces/:namespace/pipelines/:pipeline/edges", handler.ListPipelineEdges)
 	r.GET("/namespaces/:namespace/pipelines/:pipeline/edges/:edge", handler.GetPipelineEdge)
-	r.GET("/namespaces/:namespace/pipelines/:pipeline/vertices/:vertex/metrics", handler.GetVertexMetrics)
+	r.GET("/namespaces/:namespace/pipelines/:pipeline/vertices/:vertex/metrics", handler.GetVertexMetricsAndWatermark)
```
Comment (Krithika3) [inline server/routes/routes.go:36]: maybe we should not call this handle something else

**Comment by vigith** [comment]
Can we show the full date-timestamp rather than just time?

File: examples/1-simple-pipeline.yaml [hunk 2]
```diff
@@ -3,6 +3,8 @@ kind: Pipeline
 metadata:
   name: simple-pipeline
 spec:
+  watermark:
```
Comment (whynowy) [inline examples/1-simple-pipeline.yaml:6]: I would be better to create a new file particularly with watermark turned on, and leave this one for simple use case demonstration.

Comment (Krithika3) [reply examples/1-simple-pipeline.yaml:6]: Sorry, I had merged this by mistake. Will remove it :) and add a separate one for watermark

Comment (Krithika3) [reply examples/1-simple-pipeline.yaml]: Added an example for watermark

**Comment by Krithika3** [comment]
> The placement is perfect. Can we show the full date-timestamp rather than just time?
Screen Shot 2022-07-25 at 8 46 36 PM

File: examples/1-simple-pipeline.yaml (hunk 2 above)
Comment (vigith) [reply examples/1-simple-pipeline.yaml]: let's not add watermark example yet. we need to iron few more details

File: ui/src/components/pipeline/Pipeline.tsx [hunk 3]
```diff
@@ -83,23 +85,72 @@ export function Pipeline() {
...
                 json["processingRates"]["5m"].toFixed(2);
-              processinRates.ratePerFifteenMin =
+              vertexMetrics.ratePerFifteenMin =
                 json["processingRates"]["15m"].toFixed(2);
-              vertexToRateMap.set(vertex.name, processinRates);
+              vertexToMetricsMap.set(vertex.name, vertexMetrics);
             });
         })
-      ).then(() => setVertexRate(vertexToRateMap));
+      ).then(() => setVertexMetrics(vertexToMetricsMap));
```
Comment (bbehnke) [inline ui/src/components/pipeline/Pipeline.tsx:98]: Add catch here for potential error scenario.

File: ui/src/components/pipeline/Pipeline.tsx [hunk 4]
```diff
@@ -71,9 +74,8 @@ export function Pipeline() {
     }
   }, [pipeline]);
 
-  // This useEffect is used to obtain all the rates for a given vertex in a pipeline.
-  useEffect(() => {
-    const vertexToRateMap = new Map();
+  const getMetrics = () => {
```
Comment (bbehnke) [inline ui/src/components/pipeline/Pipeline.tsx:77]: Convert to useCallback

File: examples/1-simple-pipeline.yaml (hunk 2 above)
Comment (Krithika3) [reply examples/1-simple-pipeline.yaml]: ok, removed it

File: ui/src/components/pipeline/Pipeline.tsx [hunk 5]
```diff
@@ -83,23 +85,72 @@ export function Pipeline() {
...
+
+  // This useEffect is used to obtain metrics for a given vertex in a pipeline and refreshes every 5 minutes
+  useEffect(() => {
+    getMetrics();
+    const interval = setInterval(() => {
+      getMetrics();
+    }, 300000);
+
+    return () => clearInterval(interval);
+  }, [pipeline]);
```
Comment (bbehnke) [inline ui/src/components/pipeline/Pipeline.tsx:110]: Add getMetrics as dependency. Don't want to use stale instance once its update to useCallback.

File: ui/src/components/pipeline/Pipeline.tsx [hunk 6]
```diff
@@ -83,23 +85,72 @@ export function Pipeline() {
...
+              vertexToWatermarkMap.set(vertex.name, vertexWatermark);
+            });
+        })
+      ).then(() => setVertexWatermark(vertexToWatermarkMap));
+    }
+  };
+
+  // This useEffect is used to obtain watermark for a given vertex in a pipeline and refreshes every 5 seconds
+  useEffect(() => {
+    getWatermark();
```
Comment (bbehnke) [inline ui/src/components/pipeline/Pipeline.tsx:138]: Same for getWatermaerk, useCallback and add as dependency.

File: ui/src/components/pipeline/Pipeline.tsx [hunk 7]
```diff
@@ -83,23 +85,72 @@ export function Pipeline() {
...
+              const vertexWatermark = {} as VertexWatermark;
+              vertexWatermark.isWaterMarkEnabled = json["isWatermarkEnabled"];
+              vertexWatermark.watermark = json["watermark"];
+              vertexWatermark.watermarkLocalTime = new Date(
+                json["watermark"] * 1000
+              ).toLocaleString();
+              vertexToWatermarkMap.set(vertex.name, vertexWatermark);
+            });
+        })
+      ).then(() => setVertexWatermark(vertexToWatermarkMap));
```
Comment (bbehnke) [inline ui/src/components/pipeline/Pipeline.tsx:132]: catch

File: ui/src/components/pipeline/graph/Node.css [hunk 8]
```diff
@@ -17,4 +17,24 @@
 
 .node-tooltip {
     font-size: 1.2em;
+}
+
+.node-watermark {
+    width: 50px;
```
Comment (bbehnke) [inline ui/src/components/pipeline/graph/Node.css:23]: fit-content maybe better than fixed width. Not sure if a high TPS would make it overflow or not.

**Review by bbehnke — CHANGES_REQUESTED** [review]

**Review by bbehnke — APPROVED** [review]
LGTM

**Comment by whynowy** [comment]
The rate unit is `xx/sec` not `xx/min`.

**Comment by Krithika3** [comment]
> The rate unit is `xx/sec` not `xx/min`.

Changed

**Review by whynowy — APPROVED** [review]

**Pull request merged**